viewing_party/party.py

```diff
@@ -1,23 +1,112 @@
 # ------------- WAVE 1 --------------------
 
+
 def create_movie(title, genre, rating):
-    pass
+    return (
+        {"title": title, "genre": genre, "rating": rating}
+        if title and genre and rating
+        else None
+    )
+
+
+def add_to_watched(user_data, movie):
+    user_data["watched"].append(movie)
+    return user_data
+
+
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+    return user_data
+
+
+def get_movie_from_watchlist(user_data, movie_title):
+    return next((m for m in user_data["watchlist"] if m["title"] == movie_title), None)
+
+
+def watch_movie(user_data, movie_title):
+    movie = get_movie_from_watchlist(user_data, movie_title)
+    if movie:
+        user_data["watchlist"].remove(movie)
+        return add_to_watched(user_data, movie)
+    else:
+        return user_data
+
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
 
+def get_watched_avg_rating(user_data):
+    ratings = [movie["rating"] for movie in user_data["watched"]]
+    average = sum(ratings) / len(ratings) if ratings else 0
+    return average
+
+
+def get_most_watched_genre(user_data):
+    genres = [movie["genre"] for movie in user_data["watched"]]
+    most_watched = max(set(genres), key=genres.count, default=None)
+    return most_watched
+
+
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+
+def get_unique_watched(user_data):
+    friends_watched_titles = [
+        movie["title"] for friend in user_data["friends"] for movie in friend["watched"]
+    ]

```

Comment:
This is an example where we could possibly be screwed if there are different movies with the same `"title"` key, but I do think it's a reasonable assumption to make here, so I've been allowing it.

Reply:
Another nanoscopic nitpick: As this won't dedupe the titles it does make your check of `friends_watched_titles` possibly go over more movies than you need to, causing a minor performance hit.

But in general, I've been allowing it.